README.md
*Porting to Python3.10+ is painful and the progress is moving slowly.*  
*We need more volunteers to join. PRs welcome! :joy:*

ProxyBroker
===========
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-8-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->


[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=bluet_proxybroker2&metric=alert_status)](https://sonarcloud.io/summary/new_code?id=bluet_proxybroker2)
![test result](https://github.com/bluet/proxybroker2/actions/workflows/python-test-versions.yml/badge.svg)
[![GitHub issues](https://img.shields.io/github/issues/bluet/proxybroker2)](https://github.com/bluet/proxybroker2/issues)
[![GitHub stars](https://img.shields.io/github/stars/bluet/proxybroker2)](https://github.com/bluet/proxybroker2/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/bluet/proxybroker2)](https://github.com/bluet/proxybroker2/network)
[![GitHub license](https://img.shields.io/github/license/bluet/proxybroker2)](https://github.com/bluet/proxybroker2/blob/master/LICENSE)
[![Twitter](https://img.shields.io/twitter/url?style=social&url=https%3A%2F%2Fgithub.com%2Fbluet%2Fproxybroker2)](https://twitter.com/intent/tweet?text=Wow:&url=https%3A%2F%2Fgithub.com%2Fbluet%2Fproxybroker2)

ProxyBroker is an open source tool that asynchronously finds public proxies from multiple sources and concurrently checks them.

![image](https://raw.githubusercontent.com/constverum/ProxyBroker/master/docs/source/_static/index_find_example.gif)

Features
--------

-   Finds more than 7000 working proxies from \~50 sources.
-   Support protocols: HTTP(S), SOCKS4/5. Also CONNECT method to ports 80 and 23 (SMTP).
-   Proxies may be filtered by type, anonymity level, response time, country and status in DNSBL.
-   Work as a proxy server that distributes incoming requests to external proxies. With automatic proxy rotation.
-   All proxies are checked to support Cookies and Referer (and POST requests if required).
-   Automatically removes duplicate proxies.
-   Is asynchronous.

Docker
------
Docker Hub https://hub.docker.com/r/bluet/proxybroker2

```
$ docker run --rm bluet/proxybroker2 --help
  usage: proxybroker [--max-conn MAX_CONN] [--max-tries MAX_TRIES]
                     [--timeout SECONDS] [--judge JUDGES] [--provider PROVIDERS]
                     [--verify-ssl]
                     [--log [{NOTSET,DEBUG,INFO,WARNING,ERROR,CRITICAL}]]
                     [--min-queue MINIMUM_PROXIES_IN_QUEUE]
                     [--version] [--help]
                     {find,grab,serve,update-geo} ...

  Proxy [Finder | Checker | Server]

  Commands:
    These are common commands used in various situations

    {find,grab,serve,update-geo}
      find                Find and check proxies
      grab                Find proxies without a check
      serve               Run a local proxy server
      update-geo          Download and use a detailed GeoIP database

  Options:
    --max-conn MAX_CONN   The maximum number of concurrent checks of proxies
    --max-tries MAX_TRIES
                          The maximum number of attempts to check a proxy
    --timeout SECONDS, -t SECONDS
                          Timeout of a request in seconds. The default value is
                          8 seconds
    --judge JUDGES        Urls of pages that show HTTP headers and IP address
    --provider PROVIDERS  Urls of pages where to find proxies
    --verify-ssl, -ssl    Flag indicating whether to check the SSL certificates
    --min-queue MINIMUM_PROXIES_IN_QUEUE   The minimum number of proxies in the queue for checking connectivity
    --log [{NOTSET,DEBUG,INFO,WARNING,ERROR,CRITICAL}]
                          Logging level
    --version, -v         Show program's version number and exit
    --help, -h            Show this help message and exit

  Run 'proxybroker <command> --help' for more information on a command.
  Suggestions and bug reports are greatly appreciated:
  <https://github.com/bluet/proxybroker2/issues>

```


Requirements
------------

-   Python 3.8+
-   [aiohttp](https://pypi.python.org/pypi/aiohttp)
-   [aiodns](https://pypi.python.org/pypi/aiodns)
-   [maxminddb](https://pypi.python.org/pypi/maxminddb)

Installation
------------

### Install locally

To install last stable release from pypi:
> NOT RECOMMEND. It will install the out-dated original proxybroker package, which is no longer maintained by original maintainer. [https://github.com/constverum/ProxyBroker](https://github.com/constverum/ProxyBroker/issues/195)
> We will upload the up-to-date package under new name (proxybroker2) when the support for 3.10 is ready. [https://github.com/bluet/proxybroker2/issues/89](https://github.com/bluet/proxybroker2/issues/89)

``` {.sourceCode .bash}
$ pip install proxybroker
```

To install the latest development version from GitHub:

``` {.sourceCode .bash}
$ pip install -U git+https://github.com/bluet/proxybroker2.git
```

### Use pre-built Docker image

``` {.sourceCode .bash}
$ docker pull bluet/proxybroker2
```

### Build bundled one-file executable with pyinstaller

#### Requirements
Supported Operating System: Windows, Linux, MacOS

*On UNIX-like systems (Linux / macOSX / BSD)*

Install these tools
 - upx
 - objdump (this tool is usually in the binutils package)
``` {.sourceCode .bash}
$ sudo apt install -y upx-ucl binutils # On Ubuntu / Debian
```

#### Build

```
pip install pyinstaller \
&& pip install . \
&& mkdir -p build \
&& cd build \
&& pyinstaller --onefile --name proxybroker --add-data "../proxybroker/data:data" --workpath ./tmp --distpath . --clean ../py2exe_entrypoint.py \
&& rm -rf tmp *.spec
```

The executable is now in the build directory

Usage
-----

### CLI Examples

#### Find

Find and show 10 HTTP(S) proxies from United States with the high level of anonymity:

``` {.sourceCode .bash}
$ proxybroker find --types HTTP HTTPS --lvl High --countries US --strict -l 10
```

![image](https://raw.githubusercontent.com/constverum/ProxyBroker/master/docs/source/_static/cli_find_example.gif)

#### Grab

Find and save to a file 10 US proxies (without a check):

``` {.sourceCode .bash}
$ proxybroker grab --countries US --limit 10 --outfile ./proxies.txt
```

![image](https://raw.githubusercontent.com/constverum/ProxyBroker/master/docs/source/_static/cli_grab_example.gif)

#### Serve

Run a local proxy server that distributes incoming requests to a pool of found HTTP(S) proxies with the high level of anonymity:

``` {.sourceCode .bash}
$ proxybroker serve --host 127.0.0.1 --port 8888 --types HTTP HTTPS --lvl High --min-queue 5
```

![image](https://raw.githubusercontent.com/constverum/ProxyBroker/master/docs/source/_static/cli_serve_example.gif)

Run `proxybroker --help` for more information on the options available.
Run `proxybroker <command> --help` for more information on a command.

### Basic code example

Find and show 10 working HTTP(S) proxies:

``` {.sourceCode .python}
import asyncio
from proxybroker import Broker

async def show(proxies):
    while True:
        proxy = await proxies.get()
        if proxy is None: break
        print('Found proxy: %s' % proxy)

proxies = asyncio.Queue()
broker = Broker(proxies)
tasks = asyncio.gather(
    broker.find(types=['HTTP', 'HTTPS'], limit=10),
    show(proxies))

loop = asyncio.get_event_loop()
loop.run_until_complete(tasks)
```

[More examples](https://proxybroker.readthedocs.io/en/latest/examples.html).

### Proxy information per requests
#### HTTP
Check `X-Proxy-Info` header in response.
```
$ http_proxy=http://127.0.0.1:8888 https_proxy=http://127.0.0.1:8888 curl -v http://httpbin.org/get
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8888 (#0)
> GET http://httpbin.org/get HTTP/1.1
> Host: httpbin.org
> User-Agent: curl/7.58.0
> Accept: */*
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 OK
< X-Proxy-Info: 174.138.42.112:8080
< Date: Mon, 04 May 2020 03:39:40 GMT
< Content-Type: application/json
< Content-Length: 304
< Server: gunicorn/19.9.0
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< X-Cache: MISS from ADM-MANAGER
< X-Cache-Lookup: MISS from ADM-MANAGER:880
< Connection: keep-alive
<
{
  "args": {},
  "headers": {
    "Accept": "*/*",
    "Cache-Control": "max-age=259200",
    "Host": "httpbin.org",
    "User-Agent": "curl/7.58.0",
    "X-Amzn-Trace-Id": "Root=1-5eaf8e7c-6a1162a1387a1743a49063f4"
  },
  "origin": "...",
  "url": "http://httpbin.org/get"
}
* Connection #0 to host 127.0.0.1 left intact
```

#### HTTPS
We are not able to modify HTTPS traffic to inject custom header once they start being encrypted. A `X-Proxy-Info` will be sent to client after `HTTP/1.1 200 Connection established` but not sure how clients can read it.
```
(env) bluet@ocisly:~/workspace/proxybroker2$ http_proxy=http://127.0.0.1:8888 https_proxy=http://127.0.0.1:8888 curl -v https://httpbin.org/get
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8888 (#0)
* allocate connect buffer!
* Establish HTTP proxy tunnel to httpbin.org:443
> CONNECT httpbin.org:443 HTTP/1.1
> Host: httpbin.org:443
> User-Agent: curl/7.58.0
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 Connection established
< X-Proxy-Info: 207.148.22.139:8080
<
* Proxy replied 200 to CONNECT request
* CONNECT phase completed!
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
...
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x5560b2e93580)
> GET /get HTTP/2
> Host: httpbin.org
> User-Agent: curl/7.58.0
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 200
< date: Mon, 04 May 2020 03:39:35 GMT
< content-type: application/json
< content-length: 256
< server: gunicorn/19.9.0
< access-control-allow-origin: *
< access-control-allow-credentials: true
<
{
  "args": {},
  "headers": {
    "Accept": "*/*",
    "Host": "httpbin.org",
    "User-Agent": "curl/7.58.0",
    "X-Amzn-Trace-Id": "Root=1-5eaf8e77-efcb353b0983ad6a90f8bdcd"
  },
  "origin": "...",
  "url": "https://httpbin.org/get"
}
* Connection #0 to host 127.0.0.1 left intact
```

### HTTP API
#### Get info of proxy been used for retrieving specific url
For HTTP, it's easy.
```
$ http_proxy=http://127.0.0.1:8888 https_proxy=http://127.0.0.1:8888 curl -v http://proxycontrol/api/history/url:http://httpbin.org/get
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8888 (#0)
> GET http://proxycontrol/api/history/url:http://httpbin.org/get HTTP/1.1
> Host: proxycontrol
> User-Agent: curl/7.58.0
> Accept: */*
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 34
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
<
{"proxy": "..."}
```

For HTTPS, we're not able to know encrypted payload (request), so only hostname can be used.
```
$ http_proxy=http://127.0.0.1:8888 https_proxy=http://127.0.0.1:8888 curl -v http://proxycontrol/api/history/url:httpbin.org:443
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8888 (#0)
> GET http://proxycontrol/api/history/url:httpbin.org:443 HTTP/1.1
> Host: proxycontrol
> User-Agent: curl/7.58.0
> Accept: */*
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 34
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
<
{"proxy": "..."}
* Connection #0 to host 127.0.0.1 left intact
```

#### Remove specific proxy from queue
```
$ http_proxy=http://127.0.0.1:8888 https_proxy=http://127.0.0.1:8888 curl -v http://proxycontrol/api/remove/PROXY_IP:PROXY_PORT
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8888 (#0)
> GET http://proxycontrol/api/remove/... HTTP/1.1
> Host: proxycontrol
> User-Agent: curl/7.58.0
> Accept: */*
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 204 No Content
<
* Connection #0 to host 127.0.0.1 left intact
```

Documentation
-------------

<https://proxybroker.readthedocs.io/>

TODO
----

-   Check the ping, response time and speed of data transfer
-   Check site access (Google, Twitter, etc) and even your own custom URL's
-   Information about uptime
-   Checksum of data returned
-   Support for proxy authentication
-   Finding outgoing IP for cascading proxy
-   The ability to specify the address of the proxy without port (try to connect on defaulted ports)

Contributing
------------

-   Fork it: <https://github.com/bluet/proxybroker2/fork>
-   Create your feature branch: `git checkout -b my-new-feature`
-   We use [Poetry](https://python-poetry.org/) to manage dependencies. If need, install dependencies: `poetry install`
-   Commit your changes: `git commit -am 'Add some feature'`
-   Push to the branch: `git push origin my-new-feature`
-   Submit a pull request!
-   [Contributor workflow](https://github.com/bluet/proxybroker2/issues/93)

License
-------

Licensed under the Apache License, Version 2.0

*This product includes GeoLite2 data created by MaxMind, available from* [<http://www.maxmind.com>](http://www.maxmind.com).

Refs
----

-   <https://github.com/constverum/ProxyBroker/pull/161>

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/a5r0n"><img src="https://avatars.githubusercontent.com/u/32464596?v=4?s=100" width="100px;" alt=""/><br /><sub><b>a5r0n</b></sub></a><br /><a href="https://github.com/bluet/proxybroker2/commits?author=a5r0n" title="Code">💻</a></td>
    <td align="center"><a href="https://afun.tw"><img src="https://avatars.githubusercontent.com/u/4820492?v=4?s=100" width="100px;" alt=""/><br /><sub><b>C.M. Yang</b></sub></a><br /><a href="https://github.com/bluet/proxybroker2/commits?author=afunTW" title="Code">💻</a> <a href="#ideas-afunTW" title="Ideas, Planning, & Feedback">🤔</a> <a href="https://github.com/bluet/proxybroker2/pulls?q=is%3Apr+reviewed-by%3AafunTW" title="Reviewed Pull Requests">👀</a></td>
    <td align="center"><a href="http://ivanvillareal.com"><img src="https://avatars.githubusercontent.com/u/190708?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Ivan Villareal</b></sub></a><br /><a href="https://github.com/bluet/proxybroker2/commits?author=ivaano" title="Code">💻</a></td>
    <td align="center"><a href="https://github.com/quancore"><img src="https://avatars.githubusercontent.com/u/15036825?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Quancore</b></sub></a><br /><a href="https://github.com/bluet/proxybroker2/commits?author=quancore" title="Code">💻</a></td>
    <td align="center"><a href="https://github.com/synchronizing"><img src="https://avatars.githubusercontent.com/u/2829082?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Felipe</b></sub></a><br /><a href="#ideas-synchronizing" title="Ideas, Planning, & Feedback">🤔</a></td>
    <td align="center"><a href="https://www.vincentinttsh.tw/"><img src="https://avatars.githubusercontent.com/u/14941597?v=4?s=100" width="100px;" alt=""/><br /><sub><b>vincentinttsh</b></sub></a><br /><a href="https://github.com/bluet/proxybroker2/commits?author=vincentinttsh" title="Code">💻</a> <a href="https://github.com/bluet/proxybroker2/pulls?q=is%3Apr+reviewed-by%3Avincentinttsh" title="Reviewed Pull Requests">👀</a></td>
    <td align="center"><a href="https://github.com/ziloka"><img src="https://avatars.githubusercontent.com/u/50429450?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Ziloka</b></sub></a><br /><a href="https://github.com/bluet/proxybroker2/commits?author=ziloka" title="Code">💻</a></td>
  </tr>
  <tr>
    <td align="center"><a href="http://hhming.moe"><img src="https://avatars.githubusercontent.com/u/43672033?v=4?s=100" width="100px;" alt=""/><br /><sub><b>hms5232</b></sub></a><br /><a href="https://github.com/bluet/proxybroker2/commits?author=hms5232" title="Code">💻</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.rst
In compliance with the `APACHE-2.0 <https://opensource.org/licenses/Apache-2.0>`_ license: I declare that this version of the program contains my modifications, which can be seen through the usual "git" mechanism.

Change Log
==========
`Unreleased`_ (2020-07)
-----------------------
Author(s): BlueT - Matthew Lien - 練喆明, afunTW  

* Merge pull request #4 from afunTW/masterWIP: ADD: using pre-commit for code quality  
* update: end of file fixer  
* bugfix: installation error due to renaming README.rst to README.md  


`0.3.2`_ (2018-03-12)
---------------------
* Update dependencies and minor fixes
* Improve formating (by ``black``)


`0.3.1`_ (2018-02-22)
---------------------
* Update dependencies and minor fixes


`0.3.0`_ (2017-10-31)
---------------------

* Added ``update-geo`` command; This command will update the GeoIP database. After that ProxyBroker will have access to additional geolocation information of the proxy (ISO and name of region, city name) `#26`_
* Added ``--format`` flag, which indicating in what format the results will be presented
* Improved ``--outfile`` flag behavior. Earlier it was required to wait until whole process has finished, now the result is available in real-time `#39`_
* Improved way to get the external IP address `#25`_


`0.2.0`_ (2017-09-17)
---------------------

* Added CLI interface
* Added :meth:`Broker.serve` function.
  Now ProxyBroker can work as a proxy server that distributes incoming requests to a pool of found proxies
* To available types (protocols) added:
    * ``CONNECT:80`` - CONNECT method to port 80
    * ``CONNECT:25`` - CONNECT method to port 25 (SMTP)
* Added new options of checking and filtering proxies.
  :meth:`Broker.find` method has takes new parameters:
  :attr:`post`, :attr:`strict`, :attr:`dnsbl`.
  See documentation for more information
* Added check proxies to support Cookies and Referer
* Added gzip and deflate support
* :class:`Broker` attributes :attr:`max_concurrent_conn` and :attr:`attempts_conn`
  are deprecated, use :attr:`max_conn` and :attr:`max_tries` instead.
* Parameter :attr:`full` in :meth:`Broker.show_stats` is deprecated, use :attr:`verbose` instead
* Parameter :attr:`types` in :meth:`Broker.find` (and :meth:`Broker.serve`) from now is required
* :class:`ProxyChecker` renamed to :class:`Checker`.
  :class:`ProxyChecker` class is deprecated, use :class:`Checker` instead
* :attr:`Proxy.avgRespTime` renamed to :attr:`Proxy.avg_resp_time`.
  :attr:`Proxy.avgRespTime` is deprecated, use :attr:`Proxy.avg_resp_time` instead
* Improved documentation
* Major refactoring


`0.1.4`_ (2016-04-07)
---------------------

* Fixed bug when launched the second time to find proxies `#7`_


`0.1.3`_ (2016-03-26)
---------------------

* ``ProxyProvider`` renamed to ``Provider``.
  ``ProxyProvider`` class is deprecated, use ``Provider`` instead.
* ``Broker`` now accepts a list of providers and judges not only as strings
  but also objects of classes ``Provider`` and ``Judge``
* Fixed bug with signal handler on Windows `#4`_


`0.1.2`_ (2016-02-27)
---------------------

* Fixed bug with SIGINT on Linux
* Fixed bug with clearing the queue of proxy check.


`0.1`_ (2016-02-23)
-------------------

* Updated and added a few new providers
* Few minor fix


`0.1b4`_ (2016-01-21)
---------------------

* Added a few tests
* Update documentation


`0.1b3`_ (2016-01-16)
---------------------

* Few minor fix


`0.1b2`_ (2016-01-10)
---------------------

* Few minor fix


`0.1b1`_ (2015-12-29)
---------------------

* Changed the name of a PyProxyChecker on ProxyBroker in
  connection with the expansion of the concept.
* Added support of multiple proxy providers.
* Initial public release on PyPi
* Many improvements and bug fixes


`0.1a2`_ (2015-11-24)
---------------------

* Added support of multiple proxy judges.


`0.1a1`_ (2015-11-11)
---------------------

* Initial commit with function of proxy checking


.. _#4: https://github.com/constverum/ProxyBroker/issues/4
.. _#7: https://github.com/constverum/ProxyBroker/issues/7
.. _#25: https://github.com/constverum/ProxyBroker/issues/25
.. _#26: https://github.com/constverum/ProxyBroker/issues/26
.. _#39: https://github.com/constverum/ProxyBroker/issues/39
.. _0.1a1: https://github.com/constverum/ProxyBroker/compare/cf465b3
.. _0.1a2: https://github.com/constverum/ProxyBroker/compare/cf465b3...f8e2428
.. _0.1b1: https://github.com/constverum/ProxyBroker/compare/f8e2428...162f261
.. _0.1b2: https://github.com/constverum/ProxyBroker/compare/162f261...1fa10df
.. _0.1b3: https://github.com/constverum/ProxyBroker/compare/1fa10df...8f69ebd
.. _0.1b4: https://github.com/constverum/ProxyBroker/compare/8f69ebd...v0.1b4
.. _0.1: https://github.com/constverum/ProxyBroker/compare/v0.1b4...v0.1
.. _0.1.2: https://github.com/constverum/ProxyBroker/compare/v0.1...v0.1.2
.. _0.1.3: https://github.com/constverum/ProxyBroker/compare/v0.1.2...v0.1.3
.. _0.1.4: https://github.com/constverum/ProxyBroker/compare/v0.1.3...v0.1.4
.. _0.2.0: https://github.com/constverum/ProxyBroker/compare/v0.1.4...v0.2.0
.. _0.3.0: https://github.com/constverum/ProxyBroker/compare/v0.2.0...v0.3.0
.. _0.3.1: https://github.com/constverum/ProxyBroker/compare/v0.3.0...v0.3.1
.. _0.3.2: https://github.com/constverum/ProxyBroker/compare/v0.3.1...v0.3.2
.. _0.4.0: https://github.com/constverum/ProxyBroker/compare/v0.3.2...HEAD


cli.py
"""CLI."""

import argparse
import asyncio
import json
import logging
import sys
from contextlib import contextmanager

from . import __version__ as version
from .api import Broker
from .utils import update_geoip_db


def create_parser():
    parser = argparse.ArgumentParser(
        prog='proxybroker',
        add_help=False,
        description='Proxy [Finder | Checker | Server]',
        epilog='''Run '%(prog)s <command> --help'
                  for more information on a command.
                  Suggestions and bug reports are greatly appreciated:
                  https://github.com/constverum/ProxyBroker/issues''',
    )

    subparsers = parser.add_subparsers(
        dest='command',
        title='Commands',
        description='These are common commands used in various situations',
    )
    parser_group = parser.add_argument_group(title='Options')
    add_broker_args(parser_group)
    add_help_arg(parser_group)

    fparser = subparsers.add_parser(
        'find',
        add_help=False,
        help='Find and check proxies',
        description='Find and check proxies with specified parameters',
    )
    fparser_group = fparser.add_argument_group(title='Options')
    add_find_args(fparser_group)
    add_grab_args(fparser_group)
    add_limit_arg(fparser_group)
    add_outfile_arg(fparser_group)
    add_format_arg(fparser_group)
    add_show_stats_arg(fparser_group)
    add_help_arg(fparser_group)

    gparser = subparsers.add_parser(
        'grab',
        add_help=False,
        help='Find proxies without a check',
        description='Find proxies without a check with specified parameters',
    )
    gparser_group = gparser.add_argument_group(title='Options')
    add_grab_args(gparser_group)
    add_limit_arg(gparser_group)
    add_outfile_arg(gparser_group)
    add_format_arg(gparser_group)
    add_show_stats_arg(gparser_group)
    add_help_arg(gparser_group)

    sparser = subparsers.add_parser(
        'serve',
        add_help=False,
        help='Run a local proxy server',
        description='''Run a local proxy server that distributes requests to
                       external proxies, which will be found on the
                       specified parameters''',
    )
    add_serve_args(sparser.add_argument_group(title='Server options'))
    sparser_fgroup = sparser.add_argument_group(title='Find proxies options')
    add_find_args(sparser_fgroup)
    add_grab_args(sparser_fgroup)
    add_limit_arg(
        sparser_fgroup,
        100,
        '''
        When will be found a requested number of working proxies,
        checking of new proxies will be lazily paused.
        See the documentation for more information''',
    )
    add_help_arg(sparser.add_argument_group(title='Common options'))

    uparser = subparsers.add_parser(
        'update-geo',
        add_help=False,
        help='Download and use a detailed GeoIP database',
        description=(
            'Download and use a detailed GeoIP DB to get '
            'additional geolocation information of the proxy '
            '(ISO and name of region, city name).'
        ),
    )
    uparser_group = uparser.add_argument_group(title='Options')
    uparser.set_defaults(func=update_geoip_db)
    add_help_arg(uparser_group)

    return parser


def add_broker_args(group):
    group.add_argument(
        '--max-conn',
        type=int,
        default=200,
        dest='max_conn',
        help='The maximum number of concurrent checks of proxies',
    )
    group.add_argument(
        '--max-tries',
        type=int,
        default=3,
        dest='max_tries',
        help='The maximum number of attempts to check a proxy',
    )
    group.add_argument(
        '--timeout',
        '-t',
        type=int,
        default=8,
        metavar='SECONDS',
        help='''Timeout of a request in seconds.
                The default value is 8 seconds''',
    )
    group.add_argument(
        '--judge',
        action='append',
        dest='judges',
        help='Urls of pages that show HTTP headers and IP address',
    )
    group.add_argument(
        '--provider',
        action='append',
        dest='providers',
        help='Urls of pages where to find proxies',
    )
    group.add_argument(
        '--verify-ssl',
        '-ssl',
        dest='verify_ssl',
        action='store_true',
        help='Flag indicating whether to check the SSL certificates',
    )
    group.add_argument(
        '--log',
        nargs='?',
        default=logging.CRITICAL,
        choices=['NOTSET', 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'],
        help='Logging level',
    )
    group.add_argument(
        '--version',
        '-v',
        action='version',
        version='%(prog)s {v}'.format(v=version),
        help='Show program\'s version number and exit',
    )


def add_find_args(group):
    group.add_argument(
        '--types',
        nargs='+',
        type=str.upper,
        required=True,
        choices=[
            'HTTP',
            'HTTPS',
            'SOCKS4',
            'SOCKS5',
            'CONNECT:80',
            'CONNECT:25',
        ],
        help='Type(s) (protocols) that need to be check on support by proxy',
    )
    group.add_argument(
        '--lvl',
        dest='anon_lvl',
        nargs='+',
        type=str.title,
        choices=['Transparent', 'Anonymous', 'High'],
        help='Level(s) of anonymity (for HTTP only). By default, any level',
    )
    group.add_argument(
        '--data',
        type=argparse.FileType('r'),
        help='''Path to the file with proxies.
                If specified, used instead of providers''',
    )
    group.add_argument(
        '--dnsbl', nargs='+', help='Spam databases for proxy checking'
    )
    group.add_argument(
        '--post',
        action='store_true',
        help='''Flag indicating use POST instead of GET
                for requests when checking proxies''',
    )
    group.add_argument(
        '--strict',
        '-s',
        action='store_true',
        help='''Flag indicating that anonymity levels of the
                types (protocols) supported by a proxy must
                be equal to the requested types and levels of anonymity''',
    )


def add_grab_args(group):
    group.add_argument(
        '--countries',
        '-c',
        nargs='+',
        help='List of ISO country codes where should be located proxies',
    )


def add_serve_args(group):
    group.add_argument(
        '--host',
        type=str,
        default='127.0.0.1',
        help='Host of local proxy server',
    )
    group.add_argument(
        '--port', type=int, default=8888, help='Port of local proxy server'
    )
    group.add_argument(
        '--max-tries',
        type=int,
        dest='srv_max_tries',
        help='''The maximum number of attempts to handle an incoming request.
                If not specified, will be used the value passed to the %(prog)s
                command''',
    )
    group.add_argument(
        '--min-req-proxy',
        type=int,
        default=5,
        dest='min_req_proxy',
        help='''The minimum number of processed requests to decide
                whether to use it further or reject''',
    )
    group.add_argument(
        '--max-error-rate',
        type=float,
        default=0.5,
        dest='max_error_rate',
        help='''The maximum percentage of requests that ended
                with an error. For example: 0.5 = 50%%''',
    )
    group.add_argument(
        '--max-resp-time',
        type=int,
        default=8,
        dest='max_resp_time',
        metavar='SECONDS',
        help='''The maximum response time in seconds. If proxy.avg_resp_time exceeds
                this value, proxy will be rejected.
                The default value is 8 seconds''',
    )
    group.add_argument(
        '--prefer-connect',
        action='store_true',
        dest='prefer_connect',
        help='''Flag that indicates whether to use
                the CONNECT method if possible''',
    )
    group.add_argument(
        '--http-allowed-codes',
        nargs='+',
        type=int,
        dest='http_allowed_codes',
        help='Acceptable HTTP codes returned by proxy on requests',
    )
    group.add_argument(
        '--backlog',
        type=int,
        default=100,
        help='The maximum number of queued connections passed to listen',
    )


def add_limit_arg(group, _def=0, _help='The maximum number of working proxies'):
    group.add_argument('--limit', '-l', type=int, default=_def, help=_help)


def add_outfile_arg(group):
    group.add_argument(
        '--outfile',
        '-o',
        type=argparse.FileType('w', 1),
        default=sys.stdout,
        help='Save found proxies to file. By default, output to console',
    )


def add_format_arg(group):
    group.add_argument(
        '--format',
        '-f',
        nargs='?',
        type=str.lower,
        help='''Flag indicating in what format the results will be presented.
                Available formats: default and json''',
    )


def add_show_stats_arg(group):
    group.add_argument(
        '--show-stats',
        dest='show_stats',
        action='store_true',
        help='Flag indicating whether to print verbose stats',
    )


def add_help_arg(group):
    group.add_argument(
        '--help', '-h', action='help', help='Show this help message and exit'
    )


@contextmanager
def outformat(outfile, format):
    is_json = format == 'json'
    if is_json:
        outfile.write('[\n')
    try:
        yield
    finally:
        if is_json:
            outfile.write('\n]')


async def handle(proxies, outfile, format):
    with outformat(outfile, format):
        is_json = format == 'json'
        is_first = True
        while True:
            proxy = await proxies.get()
            if proxy is None:
                break

            if is_json:
                line = '%s' % json.dumps(proxy.as_json())
            else:
                line = '%r\n' % proxy

            if is_json and not is_first:
                outfile.write(',\n')
            outfile.write(line)
            is_first = False


def cli(args=sys.argv[1:]):
    parser = create_parser()
    ns = parser.parse_args(args)

    if not ns.command:
        parser.print_help()
        return
    elif ns.command == 'update-geo':
        ns.func()
        return

    logging.basicConfig(
        format='%(asctime)s - %(levelname)s - %(name)s - %(message)s',
        datefmt='[%H:%M:%S]',
        level=ns.log,
    )

    if hasattr(ns, 'anon_lvl') and 'HTTP' in ns.types:
        ns.types.remove('HTTP')
        ns.types.append(('HTTP', ns.anon_lvl))

    loop = asyncio.get_event_loop()
    proxies = asyncio.Queue(loop=loop)
    broker = Broker(
        proxies,
        max_conn=ns.max_conn,
        max_tries=ns.max_tries,
        timeout=ns.timeout,
        judges=ns.judges,
        providers=ns.providers,
        verify_ssl=ns.verify_ssl,
        loop=loop,
    )

    if ns.command in ('find', 'grab'):
        tasks = [handle(proxies, outfile=ns.outfile, format=ns.format)]
    else:
        tasks = []

    if ns.command == 'find':
        tasks.append(
            broker.find(
                data=ns.data,
                types=ns.types,
                countries=ns.countries,
                post=ns.post,
                strict=ns.strict,
                dnsbl=ns.dnsbl,
                limit=ns.limit,
            )
        )
    elif ns.command == 'grab':
        tasks.append(broker.grab(countries=ns.countries, limit=ns.limit))
    elif ns.command == 'serve':
        broker.serve(
            host=ns.host,
            port=ns.port,
            limit=ns.limit,
            min_req_proxy=ns.min_req_proxy,
            max_error_rate=ns.max_error_rate,
            max_resp_time=ns.max_resp_time,
            prefer_connect=ns.prefer_connect,
            http_allowed_codes=ns.http_allowed_codes,
            backlog=ns.backlog,
            data=ns.data,
            types=ns.types,
            countries=ns.countries,
            post=ns.post,
            strict=ns.strict,
            dnsbl=ns.dnsbl,
        )
        print('Server started at http://%s:%d' % (ns.host, ns.port))

    try:
        if tasks:
            loop.run_until_complete(asyncio.gather(*tasks, loop=loop))
            if ns.show_stats:
                broker.show_stats(verbose=True)
        else:
            loop.run_forever()
    except KeyboardInterrupt:
        broker.stop()
    finally:
        loop.stop()
        loop.close()
